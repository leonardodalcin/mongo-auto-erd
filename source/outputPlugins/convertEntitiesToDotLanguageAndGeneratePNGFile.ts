import { IEntity } from '@interfaces/IEntity'
import { digraph } from 'graphviz'
import { resolve } from 'path'

function entityToNodeLabel(entity: IEntity) {
  const propLabels = entity.properties.map((property) => {
    return `${property.name} : ${property.types[0]}`
  })

  const label = `{${entity.name} | ${
    propLabels.length > 0
      ? propLabels.reduce((previousValue = '', currentValue) => {
          return `${previousValue}\\l${currentValue}`
        })
      : ''
  }}`
  return label
}

export function convertEntitiesToDotLanguageAndGeneratePNGFile(
  entities: IEntity[],
  filepath?: string
) {
  const graph = digraph('erd')
  for (const entity of entities) {
    graph.addNode(entity.name, {
      shape: 'record',
      label: entityToNodeLabel(entity)
    })
  }
  for (const entity of entities) {
    for (const relationship of entity.relationships) {
      if (entity.name !== relationship.targetCollectionName)
        graph.addEdge(entity.name, relationship.targetCollectionName)
    }
  }

  if (filepath) graph.output('png', resolve(filepath))

  return graph.to_dot()
}
