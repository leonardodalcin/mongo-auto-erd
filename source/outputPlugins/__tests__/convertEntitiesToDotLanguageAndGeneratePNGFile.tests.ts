import { convertEntitiesToDotLanguageAndGeneratePNGFile } from '@outputPlugins/convertEntitiesToDotLanguageAndGeneratePNGFile'
import { ERD } from './ERD.mock'

describe('convertEntitiesToDotLanguageAndGeneratePNGFile.', () => {
  it('should convert mock to a valid dot file', () => {
    const dotLanguageGraph = convertEntitiesToDotLanguageAndGeneratePNGFile(ERD)
    expect(dotLanguageGraph).toMatchSnapshot()
  })
})
