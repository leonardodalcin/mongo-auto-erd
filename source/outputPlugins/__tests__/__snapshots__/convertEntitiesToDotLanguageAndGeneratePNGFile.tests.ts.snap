// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertEntitiesToDotLanguageAndGeneratePNGFile. should convert mock to a valid dot file 1`] = `
"digraph erd {
  \\"users\\" [ label = \\"{users | __v : number\\\\l_id : objectId\\\\lname : string}\\", shape =record ];
  \\"photos\\" [ label = \\"{photos | __v : number\\\\l_id : objectId\\\\ltitle : string\\\\luser : objectId}\\", shape =record ];
  \\"usergroups\\" [ label = \\"{usergroups | __v : number\\\\l_id : objectId\\\\lname : string\\\\lusers : objectId}\\", shape =record ];
  \\"photos\\" -> \\"users\\";
  \\"usergroups\\" -> \\"users\\";
}
"
`;
